[Documentation]: Add more info about contextual patterns and pattern transformations (#38809)
WordPress/gutenberg: docs/reference-guides/block-api/block-patterns.md

The editor comes with several core block patterns. Theme and plugin authors can register additional custom block patterns using the `register_block_pattern` helper function.

> The `register_block_pattern` helper function receives two arguments.

Author: There are some auto formatting changes that removed some extra spaces like this one. The main changes are in `blockTypes` and `contextual patterns` sections

-   `categories` (optional): An array of registered pattern categories used to group block patterns. Block patterns can be shown on multiple categories. A category must be registered separately in order to be used here.
-   `keywords` (optional): An array of aliases or keywords that help users discover the pattern while searching.
-   `viewportWidth` (optional): An integer specifying the intended width of the pattern to allow for a scaled preview of the pattern in the inserter.

> -   `blockTypes` (optional): An array of block types that the pattern is intended to be used with. Each value needs to be the declared block's `name`. This is used for finding proper matches in the patterns transformation mechanism. In addition this is used by `__experimentalBlockPatternSetup` component which displays patterns to choose from in some blocks, like `Query Loop` block.

Reviewer: > This is used for finding proper matches in the patterns transformation mechanism. In addition this is used by `__experimentalBlockPatternSetup` component which displays patterns to choose from in some blocks, like `Query Loop` block.

Personally I'd drop this sentence from here in favor of the explications later in this file. Because it's more complex than that. If we choose to keep I'd remove the mention of the experimental component and just say "setup state"

unregister_block_pattern_category( 'hero' );
```

_Note:_

`unregister_block_pattern_category()` should be called from a handler attached to the init hook.

```php
function my_plugin_unregister_my_pattern_categories() {
  unregister_block_pattern_category( ... );
}

add_action( 'init', 'my_plugin_unregister_my_pattern_categories' );
```

## Block patterns contextual to block types and pattern transformations

It is possible to attach a block pattern to one or more block types. This adds the block pattern as an available transform for that block type.

> Currently pattern transformations are available for simple blocks(blocks without inner blocks). In order for a pattern to be suggested, **every selected block must have matched with a pattern's block**. If a suggested pattern has more blocks than the selected blocks, they are added `as are` from the pattern. Every pattern block that matched with a selected block will be transformed. Noting that this doesn't necessarily mean the order of the blocks will be preserved.

Reviewer: ```suggestion
Currently these transformations are available only to simple blocks (blocks without inner blocks). In order for a pattern to be suggested, **every selected block must be present in the block pattern**.
```

Personally, I'm not sure the details about how the transformation works are needed here, it felt just confusing to me here and more suited later (as the examples show). Happy if you think otherwise.
